### DISCUSSION TITLE
「RFC」New Component：ModalConfig 模型参数面板

### DISCUSSION THREAD

**Discussion by ONLY-yours** [ant-design/pro-chat#58] [RFCs]
### Summary
新增 ModalConfig 模型参数面板组件
### Motivation
现在 ProChat 只能通过 Config 直接配置进去，有些时候用户会需要一个切换面板 or 细节的参数调整面板，能够在对话过程中进行动态的调整。
> 默认会在 ProChat 中关闭，提供一个类似 ConfigPannel 的 api 可以配置，配置后会默认直接在某一侧打开改面板，修改好的 Pannel 会在 Request 的第二个参数中进行动态变化，该配置会高于默认的 Config 配置

### Example
参考如下，类似于这样子一个面板
![Sd-WebUI](https://github.com/ant-design/pro-chat/assets/52664827/1d04fdae-a6a4-4353-8d8a-4a86bf2918dd)

![ChatGLM](https://github.com/ant-design/pro-chat/assets/52664827/a761030c-53d6-48de-82a8-8b33a8a99e04)

### API （还在构想）
这块还没怎么想好，有些应该是不会开放出去的，例如 historyCount、fewShots 这类，比较犹豫的是类似于 Modal（所使用的语言模型）
目前可支持配置的模型参数为：

- 模型参数部分（和大部分 LLM 模型有关）
   - frequency_penalty：生成频率程度
   - max_tokens：最长的 token 限制
   - presence_penalty：生成主题程度
   - temperature：文本随机量
   - top_p ：生成文本最高单个 token
- systemRole：系统角色（例如：你是一个XXX助手，希望你扮演这个角色帮我干事情）
- inputTemplate（不确定）：输入模版（例如：接下来这段话我需要你帮我翻译为中文：{inputString}，并总结一下）
- modal（不确定）：模型参数（是一个对象，例如：{name:"gpt3.5",hashid:"xxxx"}）

目前的 Config 所有参数如下：
```tsx
export interface ModelConfig {
  compressThreshold?: number;
  /**
   * 历史消息长度压缩阈值
   */
  enableCompressThreshold?: boolean;
  /**
   * 开启历史记录条数
   */
  enableHistoryCount?: boolean;
  enableMaxTokens?: boolean;
  /**
   * 语言模型示例
   */
  fewShots?: LLMFewShots;
  /**
   * 历史消息条数
   */
  historyCount?: number;
  inputTemplate?: string;
  /**
   * 角色所使用的语言模型
   */
  model?: string;
  /**
   * 语言模型参数
   */
  params?: ModelParams;
  /**
   * 系统角色
   */
  systemRole?: string;
}


// 语言模型的设置参数
export interface ModelParams {
  /**
   * 控制生成文本中的惩罚系数，用于减少重复性
   * @default 0
   */
  frequency_penalty?: number;
  /**
   * 生成文本的最大长度
   */
  max_tokens?: number;
  /**
   * 控制生成文本中的惩罚系数，用于减少主题的变化
   * @default 0
   */
  presence_penalty?: number;
  /**
   * 生成文本的随机度量，用于控制文本的创造性和多样性
   * @default 0.6
   */
  temperature?: number;
  /**
   * 控制生成文本中最高概率的单个 token
   * @default 1
   */
  top_p?: number;

  [key: string]: any;
}

export type ModelRoleType = 'user' | 'system' | 'assistant' | 'function';

export interface LLMMessage {
  content: string;
  role: ModelRoleType;
}

export type LLMFewShots = LLMMessage[];

```

### Some Design
搜集一些别人做的好的样例

DIfy：[https://dify.ai/](https://dify.ai/)
![image](https://github.com/ant-design/pro-chat/assets/52664827/e14b2710-575d-45c8-adf6-48c1824ccb9b)

Coze：[https://www.coze.com/](https://www.coze.com/)
![image](https://github.com/ant-design/pro-chat/assets/52664827/710bdbc0-e4a0-4a93-8594-d71acac46705)

欢迎提供更多好的想法输入

**Comment by KrishnaPG** [comment]
I think it is better to use the [Formily](https://react.formilyjs.org/guide) approach. You supply the config and it will render the form dynamically on the fly. This allows:

1. Flexibility for user to have their own form render if they wants to (the default one being dynamic formily form)
2. When the config changes in the later versions, no need to redesign the forms, since they are rendered on the fly dynamically.

In the ProChat input component UI, allow a button somewhere which when clicked will invoke the Formily (which renders the current config dynamically as a form).
